pkg/secrethub/credentials/readers.go

```diff
@@ -1,52 +1,98 @@
 package credentials
 
 import (
+	"fmt"
 	"io/ioutil"
 	"os"
 )
 
-// Reader helps with reading bytes from a configured source.
-type Reader interface {
+type ErrDecodingCredential struct {
+	Location string
+	Err      error
+}
+
+func (e ErrDecodingCredential) Error() string {
+	return fmt.Sprintf("error decoding credential loaded from '%s': %v", e.Location, e.Err)
+}
+
+// PassphraseReader helps with reading bytes from a configured source.
+type PassphraseReader interface {
 	// Read reads from the reader and returns the resulting bytes.
 	Read() ([]byte, error)
 }
 
+type passphraseReader func() ([]byte, error)
+
+func (p passphraseReader) Read() ([]byte, error) {
+	return p()
+}
+
+// PassphraseFromString returns a reader that simply returns the given string as
+// a byte slice when Read() is called.
+func PassphraseFromString(passphrase string) PassphraseReader {
+	return passphraseReader(func() ([]byte, error) {
+		return []byte(passphrase), nil
+	})
+}
+
+type KeyReader interface {
+	Read(decoder KeyDecoder) (Key, error)
+}
+
 // FromFile returns a reader that reads the contents of a file,
 // e.g. a credential or a passphrase.
-func FromFile(path string) Reader {
-	return readerFunc(func() ([]byte, error) {
-		return ioutil.ReadFile(path)
+func FromFile(path string) KeyReader {
+	return keyReaderFunc(func(decoder KeyDecoder) (Key, error) {
+		keyBytes, err := ioutil.ReadFile(path)
+		if err != nil {
+			return Key{}, err
+		}
+		key, err := decoder.Decode(keyBytes)
+		if err != nil {
+			return Key{}, ErrDecodingCredential{
+				Location: path,
+				Err:      err,
+			}
+		}
+		return key, nil
 	})
 }
 
 // FromEnv returns a reader that reads the contents of an
 // environment variable, e.g. a credential or a passphrase.
-func FromEnv(key string) Reader {
-	return readerFunc(func() ([]byte, error) {
-		return []byte(os.Getenv(key)), nil
+func FromEnv(envVarKey string) KeyReader {
+	return keyReaderFunc(func(decoder KeyDecoder) (Key, error) {
+		key, err := decoder.Decode([]byte(os.Getenv(envVarKey)))
+		if err != nil {
+			return Key{}, ErrDecodingCredential{
+				Location: "$" + envVarKey,
+				Err:      err,
+			}
+		}
+		return key, nil
 	})
 }
```

Comment:
In both of these functions we need to decide whether or not to delay the reading of the credential (file/envvar).